Highlight identified features in 3d views

uclaros opened the pull request:
## Description
This is a feature promised in our [latest crowdfund campaign](https://www.lutraconsulting.co.uk/crowdfunding/qgis-3d-for-open-source-digital-twins).

The identified polygon features are now highlighted in 3D views using a semi transparent overlay and a thick silhouette line, similar to the `QgsHighlight` effect of the 2D map canvas, with a matching color and transparency. 

<img width="1091" height="579" alt="image" src="https://github.com/user-attachments/assets/08136c57-43e1-41e8-8003-0bb35045037a" />

Rendering is handled in a new render view, consisting of two passes, one for the semi transparent highlight, while writing to the stencil buffer, and one using 4 displaced viewports while reading the stencil buffer to achieve the silhouette effect.

The highlight effect ignores the depth test so that it is always rendered on top of other entities.

![Peek 2025-12-16 14-46](https://github.com/user-attachments/assets/f5946758-d5a4-4a4d-b89e-e8d53c805895)

When highlighted features overlap in the current point of view, one aggregated silhouette is rendered at the perceived perimeter.

![Peek 2025-12-16 14-50](https://github.com/user-attachments/assets/e92a86ef-11a8-4acb-abb8-2449ea0bcaff)

Funded by: https://www.lutraconsulting.co.uk/crowdfunding/qgis-3d-for-open-source-digital-twins
<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - you have read the QGIS Developer Guide (https://docs.qgis.org/testing/en/docs/developers_guide/index.html) and your PR complies with its QGIS Coding Standards
-->

uclaros committed:
Enable stencil usage in the depth buffer

uclaros committed:
Add new QgsHighlightMaterial and singlecolor fragment shader

uclaros committed:
Render identified features in 3d map canvas

uclaros committed:
Sort includes

uclaros committed:
docstrings

uclaros committed:
deleteLater over qDeleteAll

uclaros committed:
Simplify tree structure: move 4 viewports to (almost) leafs

uclaros committed:
Move highlights rendering to a render view
Handle all highlighted features in a single feature handler

uclaros committed:
fix lighting, light entities need the render QLayer

uclaros committed:
fix data defined properties

uclaros committed:
also handle rule based 3d renderers

uclaros committed:
Properly delete feature 3d handlers

uclaros committed:
Rename slot and docstring

uclaros committed:
Reproject geometries to 3d view crs if needed

benoitdm-oslandia commented:
Sometimes I wonder how this kind of situation can arise: see https://github.com/qgis/QGIS/pull/64320

uclaros committed:
cond private

uclaros committed:
virtual destructor

uclaros commented:
> Sometimes I wonder how this kind of situation can arise: see #64320

Well, I hope you're not working on other features of the crowdfunding campaign too!

uclaros committed:
remove unused vars

uclaros committed:
update framegraph existing test

ptitjano commented:
> > What a pity.
> 
> @signedav indeed... we have added feature highlighting to the list of crowdfunding topics explicitly because the Frankfurt bid had no such work item, so we expected no such clash...

We didn't initially intend to work on highlighting, but while developing the snapping feature, we realized that the user experience is significantly degraded if the geometry being snapped isn't highlighted. Indeed, when you click on a 3D scene, it can sometimes feel like you are above an object, whereas you are actually positioned next to it. Highlighting the current object helps resolve this issue.

Therefore, we developed a highlighting solution that works on all vector layers. We also realized that this solution improves the identify tool. So we ended up with #64320.

At that point, we didn't realize you were working on the same feature.

Aside from the technical aspects, ours solutions differ in terms of rendering: ours allows to stay close to the geometry (necessary for the snapping), while yours produces a visual effect that is more suitable to highlight an identified feature.

It would be a shame/counterproductive to end up with two different implementations for highlighting a vector object.

It is therefore necessary to find a compromise with one implementation that allows two highlighting modes:
- a visual effect as proposed in your PR used by the identify tool
- an opaque highlight for the measure tool: close to the geometry vertices, it also highlights the edges in a different color

Will you be able to adapt your implementation to produce these two modes so that snapping is also possible? This means to let the tool chooses if the highlight effect is a scale up or a close up of the feature geometry.

uclaros committed:
Merge branch 'master' into 3d-highlights

uclaros committed:
Merge branch 'master' into 3d-highlights

uclaros committed:
Merge branch 'master' into 3d-highlights

uclaros committed:
string literals shortcuts

uclaros committed:
Merge branch 'master' into 3d-highlights

uclaros committed:
Move the highlight material creation logic to the QgsAbstractMaterialSettings subclasses
This also allowed to add support for highlighting instanced points and 3d models

uclaros committed:
Merge branch 'master' into 3d-highlights

uclaros committed:
Use QgsBox3D for chunk extents

uclaros committed:
fix unity builds

uclaros committed:
fix unity build, remove unused includes

uclaros committed:
added highlighting rendering tests

uclaros committed:
Don't add forward and shadows QLayer to all entities in offscreen engine too.

uclaros committed:
Fix and extend tests

uclaros committed:
Cleanup includes

uclaros committed:
Merge branch 'master' into 3d-highlights

uclaros committed:
apply updated pre-commit to new files

wonder-sk commented on src/3d/framegraph/qgshighlightsrenderview.cpp:
```diff
@@ -0,0 +1,197 @@
+  Qt3DRender::QRenderTargetSelector *highlightsRenderTargetSelector = new Qt3DRender::QRenderTargetSelector( mMainViewport );
+  highlightsRenderTargetSelector->setTarget( mRenderTarget );
+
+  // First Pass:
```
```suggestion
  // Step 1: draw semi-transparent highlight
  //
```

wonder-sk commented on src/3d/materials/qgsnullmaterialsettings.cpp:
```diff
@@ -64,8 +65,11 @@ bool QgsNullMaterialSettings::equals( const QgsAbstractMaterialSettings *other )
+QgsMaterial *QgsNullMaterialSettings::toMaterial( QgsMaterialSettingsRenderingTechnique technique, const QgsMaterialContext &context ) const
 {
+  if ( context.isHighlighted() )
+    return new QgsHighlightMaterial( technique );
```
if I understand correctly, null material means we should be rendering the object with 2D map rendering to the terrain texture... so I guess we should not be using QgsHighlighMaterial here?

uclaros committed:
Address review comments

uclaros committed:
Add support for simple lines. Since they are single color anyway they don't need the QgsHighlightMaterial

uclaros committed:
clang-tidy

uclaros committed:
fix ambient occlusion test that needs to be updated upon every change to the framegraph
